node/core/pvf/src/artifacts.rs

```diff
@@ -14,6 +14,46 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
+//! PVF artifacts (final compiled code blobs).
+//!
+//!	# Lifecycle of an artifact
+//!
+//! 1. During node start-up, the artifacts cache is cleaned up.
+//!
+//! 2. In order to be executed, a PVF should be prepared first. This means that artifacts should
+//!    have an [`ArtifactState::Prepared`] entry for that artifact. If not, the preparation process
+//!    kicks in. The execution request is stashed until after the preparation is done, and the
+//!    artifact state in the host is set to [`ArtifactState::Preparing`]. Preparation goes through
+//!    the preparation queue and the pool.
+//!
+//!    1. If the artifact is already being processed, we add another execution request to the
+//!       existing preparation job, without starting a new one.
+//!
+//!    2. Note that if the state is [`ArtifactState::FailedToProcess`], we usually do not retry
```

Comment:
We do not currently have any "I'm taking a long time" messages, so if we send out approval assignments but do artifact builds lazily, then we'll cause no shows, given that builds can take more than the 12? second no show time out.

In theory, we could send messages for "building artifact" and/or "It's slow but I'm here", but @rphmeier wanted to avoid complicating the approval process with such messages, probably a wise decision.  We therefore need PVF artifacts to be built in advance, or else we suck up the risk of correlated artifact builds creating de fact escalations.

Reply:
Indeed, I was wondering about this a couple of times already myself. I think for the time being, preparation is usually pretty fast so there are no issues.

The problem with preparation in advance is, that this will likely result in wasted effort in case of parathreads. As all validators would need to prepare a PVF, although only 30 approval checkers will actually need it. Might be fine. 

Other options:

Gather some data on actual preparation times - if they turn out to be very low, we could just enforce a reasonably low value in pre-checking.

Reply:
We should imho avoid clearing the artifacts cache when the host did not change.  We could recompile only the parachain blocks when the host version did change, then lazily recompile the parathread ones.  

We should've timings of course, but we'll never stop people building wasm blobs that screw up build times intentionally.

Interpreting kinda works.  We have consensus upon who gets compiled vs interpreted, so interpreted then runs with different approval time parameters.  We could similarly adjust approval parameters to include recompiling parathreads each block.  This makes parathreads more expensive and second class though.  We could've parathreads that "buy" being compiled in advance like parachains. 

We do still have everyone compile the parathread when the PVF initially gets uploaded though, yes?  I'd think this suggests parathreads and parachains should be all be precompiled, which just makes uploading a PVF more expensive.

Reply:
> We should imho avoid clearing the artifacts cache when the host did not change.

We'd need an intelligent garbage collector, then. Imagine the node is restarted and has a hundred artifacts in the cache. How do we know which ones we will use and which are stale?

Reply:
> We should've timings of course, but we'll never stop people building wasm blobs that screw up build times intentionally.

Why? They would not pass the pre-checking phase, assuming:

> Gather some data on actual preparation times - if they turn out to be very low, we could just enforce a reasonably low value in pre-checking.

Reply:
We'd need to garbage collect PVFs when parachains deregister or when the PVF gets superseded by later PVFs.  We could do pre-checking after a PVF upload gets finalized, so then we avoid fork concerns and each parachan has at most two PVFs in the cache.  We'd loose the ability to upgrade parachain PVFs when finality stalls though, so system parachain could require some escape hatch here.

As always we pay for optimizations with complexity.  Ain't clear how far this should go right now of course.  We could stick with the current proposal for now, but make an issue for smarter PVF  garbage collection in future.

> Why? They would not pass the pre-checking phase, assuming:

It'll be possible to pass the pre-checking but be quite slow compared with average PVF builds.

It'll occasionally be possible to pass the pre-checking on one host, but be abysmal on some host upgrade in the pipeline.   We could imho ignore this risk though, so yeah maybe you're right..

Reply:
> As always we pay for optimizations with complexity. Ain't clear how far this should go right now of course. We could stick with the current proposal for now, but make an issue for smarter PVF garbage collection in future.

That does sound like it would introduce some complexity. I guess my question would be, is it really necessary? How much disk space does each compiled PVF actually require? How bad is it to keep old artifacts around? So far, it seems that we have not had issues with the 24-hour TTL of artifacts, so my grug brain thinks that we shouldn't introduce unnecessary optimizations. For parathreads I could see the artifacts needing to stay around for longer - but in that case I would just have a longer TTL for those, and not worry about the extra used disk space. 😛

Reply:
All this complexity comes from one scenario:  An adversary can create many relay chain forks, so they can upload one artifact on each fork.  We support all forks until we know which fork survives.

We've rough consensus on artifact age so we could use duration as a proxy for finality though:  We retain all artifacts compatible with the current host, so long as either the artifact is active on some relay chain fork, or else the artifact is less than 24 hours old.

We add some abandon artifact call for artifacts uploaded but not activated, or else force activation at some block height, or something like that.

It's messy to create many relay chain forks without equivocation, so the attack might already result in slashing, which maybe suffices.  If you've a run of blocks, then you could've some forks without equivocations, but not too many.  

It's maybe just easier to wait for finality and have some override for system parachains, not sure.

Reply:
Raised https://github.com/paritytech/polkadot/issues/6941 to continue this discussion.